File: 04/CSharp/QueueApp/ArrayQueue.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace QueueApp
{
    public class ArrayQueue : AbstractQueue
    {
        private Object[] array;
        private int size;
        private int capacity;
        private const int _defaultCapacity = 10;
        private int head;
        private int tail;

        public ArrayQueue(int size)
        {
            if (size < 1)
                throw new InvalidOperationException("ERROR: size < 1!");

            array = new Object[size];
            capacity = size;
            this.size = 0;
            head = -1;
            tail = 0;
        }
        public ArrayQueue()
        {
            array = new Object[_defaultCapacity];
            capacity = _defaultCapacity;
            size = 0;
            head = -1;
            tail = 0;
        }
        override public bool IsEmpty()
        {
            return size == 0;
        }
        override public void Enqueue(Object newElement)
        {
            if (size == capacity)
            {
                Object[] newQueue = new Object[2 * capacity];
                Array.Copy(array, 0, newQueue, 0, array.Length);
                array = newQueue;
                capacity *= 2;
                newQueue = null;
            }
            size++;
            array[tail++ % capacity] = newElement ?? throw new InvalidOperationException("ERROR: NULL elenent!");
        }
        override public Object Dequeue()
        {
            if (size == 0)
            {
                throw new InvalidOperationException("ERROR: Empty queue!");
            }
            size--;
            return array[++head % capacity];
        }
        override public Object Element()
        {
            if (size == 0)
            {
                throw new InvalidOperationException("ERROR: Empty queue!");
            }
            return array[head + 1 % capacity];
        }
        override public void Clear()
        {
            while (size != 0)
                Dequeue();
        }
        override public int Size()
        {
            return size;
        }
        public int Count
        {
            get
            {
                return size;
            }
        }
        public Object[] ToArray()
        {
            if (IsEmpty())
                throw new InvalidOperationException("ERROR: NULL elenent!");

            Object[] rezult = new Object[size];
            Array.Copy(array, head + 1 % capacity, rezult, 0, size);
            return rezult;
        }
    }
}

File: 04/Java/ArrayQueue.java
import java.util.function.Predicate;
import java.util.function.Function;

public class ArrayQueue extends AbstractQueue implements Queue {
    private int head;
    private int tail;
    private int size;
    private Object[] elements = new Object[10];

    private void fixCapacity(int capacity) {
        int len = elements.length;
        if (capacity > len) {
            Object[] newElements = new Object[elements.length * 2];
            int i = 0;
            while (tail!=head) {
                newElements[i] = elements[head];
                head = (head + 1) % len;
                i++;
            }
            head = 0;
            tail = len - 1;
            elements = newElements;
        }
    }

    public void enqueue(Object element) {
        assert element != null;
        fixCapacity(size + 2);
        elements[tail] = element;
        tail = (tail + 1) % elements.length;
        size++;
    }

    public Object element() {
        assert size > 0;
        return elements[head];
    }

    public Object dequeue() {
        Object ret = element();
        elements[head] = null;
        head = (head + 1) % elements.length;
        size--;
        return ret;
    }

    public void clear() {
        head = 0;
        tail = 0;
        size = 0;
    }

    public void push(Object element) {
        assert element != null;
        fixCapacity(size + 2);
        head = head - 1;
        if (head < 0) {
            head = elements.length - 1;
        }
        elements[head] = element;
        size++;
    }

    public Object peek() {
        assert size > 0;
        int ltail = tail - 1;
        if (ltail < 0) {
            ltail = elements.length - 1;
        }
        return elements[ltail];
    }
   
    public Object remove() {
        assert size > 0;
        tail = tail - 1;
        if (tail < 0) {
            tail = elements.length - 1;
        }
        Object ret = elements[tail];
        elements[tail] = null;
        size--;
        return ret;
    }

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return (size == 0);
    }

    public ArrayQueue makeCopy() {
        ArrayQueue copy = new ArrayQueue();
        copy.elements = new Object[elements.length];
        for (int i = 0; i < elements.length; i++) {
            copy.elements[i] = elements[i];
        }
        copy.size = size;
        copy.head = head;
        copy.tail = tail;
        return copy;
    }
}